<!DOCTYPE html><html lang="en" mode="light" ><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="pv-cache-enabled" content="false"><meta name="generator" content="Jekyll v4.2.0" /><meta property="og:title" content="Numpy로 구현하는 GBLUP - Imprementing GBLUP with numpy" /><meta name="author" content="Hyo-Jun Lee" /><meta property="og:locale" content="en_US" /><meta name="description" content="사용된 데이터와 코드는 아래의 url 에서 확인 하실 수 있습니다. The data and codes used in the poster can be found at https://github.com/gywns6287/Implementing-GBLUP-with-numpy" /><meta property="og:description" content="사용된 데이터와 코드는 아래의 url 에서 확인 하실 수 있습니다. The data and codes used in the poster can be found at https://github.com/gywns6287/Implementing-GBLUP-with-numpy" /><link rel="canonical" href="https://gywns6287.github.io/posts/GBLUP/" /><meta property="og:url" content="https://gywns6287.github.io/posts/GBLUP/" /><meta property="og:site_name" content="Hyo-Jun Lee" /><meta property="og:type" content="article" /><meta property="article:published_time" content="2021-02-01T14:10:00+09:00" /><meta name="twitter:card" content="summary" /><meta property="twitter:title" content="Numpy로 구현하는 GBLUP - Imprementing GBLUP with numpy" /><meta name="twitter:site" content="@twitter_username" /><meta name="twitter:creator" content="@Hyo-Jun Lee" /><meta name="google-site-verification" content="google_meta_tag_verification" /> <script type="application/ld+json"> {"author":{"@type":"Person","name":"Hyo-Jun Lee"},"description":"사용된 데이터와 코드는 아래의 url 에서 확인 하실 수 있습니다. The data and codes used in the poster can be found at https://github.com/gywns6287/Implementing-GBLUP-with-numpy","url":"https://gywns6287.github.io/posts/GBLUP/","@type":"BlogPosting","headline":"Numpy로 구현하는 GBLUP - Imprementing GBLUP with numpy","dateModified":"2021-03-03T14:41:19+09:00","datePublished":"2021-02-01T14:10:00+09:00","mainEntityOfPage":{"@type":"WebPage","@id":"https://gywns6287.github.io/posts/GBLUP/"},"@context":"https://schema.org"}</script><title>Numpy로 구현하는 GBLUP - Imprementing GBLUP with numpy | Hyo-Jun Lee</title><link rel="shortcut icon" href="/assets/img/favicons/favicon.ico" type="image/x-icon"><link rel="icon" href="/assets/img/favicons/favicon.ico" type="image/x-icon"><link rel="apple-touch-icon" href="/assets/img/favicons/apple-icon.png"><link rel="apple-touch-icon" href="/assets/img/favicons/apple-icon-precomposed.png"><link rel="apple-touch-icon" sizes="57x57" href="/assets/img/favicons/apple-icon-57x57.png"><link rel="apple-touch-icon" sizes="60x60" href="/assets/img/favicons/apple-icon-60x60.png"><link rel="apple-touch-icon" sizes="72x72" href="/assets/img/favicons/apple-icon-72x72.png"><link rel="apple-touch-icon" sizes="76x76" href="/assets/img/favicons/apple-icon-76x76.png"><link rel="apple-touch-icon" sizes="114x114" href="/assets/img/favicons/apple-icon-114x114.png"><link rel="apple-touch-icon" sizes="120x120" href="/assets/img/favicons/apple-icon-120x120.png"><link rel="apple-touch-icon" sizes="144x144" href="/assets/img/favicons/apple-icon-144x144.png"><link rel="apple-touch-icon" sizes="152x152" href="/assets/img/favicons/apple-icon-152x152.png"><link rel="apple-touch-icon" sizes="180x180" href="/assets/img/favicons/apple-icon-180x180.png"><link rel="icon" type="image/png" sizes="192x192" href="/assets/img/favicons/android-icon-192x192.png"><link rel="icon" type="image/png" sizes="32x32" href="/assets/img/favicons/favicon-32x32.png"><link rel="icon" type="image/png" sizes="96x96" href="/assets/img/favicons/favicon-96x96.png"><link rel="icon" type="image/png" sizes="16x16" href="/assets/img/favicons/favicon-16x16.png"><link rel="manifest" href="/assets/img/favicons/manifest.json"><meta name='msapplication-config' content='/assets/img/favicons/browserconfig.xml'><meta name="msapplication-TileColor" content="#ffffff"><meta name="msapplication-TileImage" content="/assets/img/favicons/ms-icon-144x144.png"><meta name="theme-color" content="#ffffff"><link rel="preconnect" href="https://fonts.gstatic.com" crossorigin="anonymous"><link rel="dns-prefetch" href="https://fonts.gstatic.com"><link rel="preconnect" href="https://www.google-analytics.com" crossorigin="use-credentials"><link rel="dns-prefetch" href="https://www.google-analytics.com"><link rel="preconnect" href="https://www.googletagmanager.com" crossorigin="anonymous"><link rel="dns-prefetch" href="https://www.googletagmanager.com"><link rel="preconnect" href="cdn.jsdelivr.net"><link rel="dns-prefetch" href="cdn.jsdelivr.net"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@4.0.0/dist/css/bootstrap.min.css" integrity="sha256-LA89z+k9fjgMKQ/kq4OO2Mrf8VltYml/VES+Rg0fh20=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@5.11.2/css/all.min.css" integrity="sha256-+N4/V/SbAFiW1MPBCXnfnP9QSN3+Keu+NlB+0ev/YKQ=" crossorigin="anonymous"><link rel="stylesheet" href="/assets/css/style.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@1.0.1/dist/bootstrap-toc.min.css"> <script src="https://cdn.jsdelivr.net/npm/jquery@3/dist/jquery.min.js"></script> <script defer src="https://cdn.jsdelivr.net/combine/npm/popper.js@1.15.0,npm/bootstrap@4/dist/js/bootstrap.min.js"></script> <script defer src="/assets/js/dist/post.min.js"></script> <script defer src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script defer src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script> <script defer src="/app.js"></script> <script defer src="https://www.googletagmanager.com/gtag/js?id=G-SLE093F45F"></script> <script> document.addEventListener("DOMContentLoaded", function(event) { window.dataLayer = window.dataLayer || []; function gtag(){dataLayer.push(arguments);} gtag('js', new Date()); gtag('config', 'G-SLE093F45F'); }); </script><body data-spy="scroll" data-target="#toc"><div id="sidebar" class="d-flex flex-column align-items-end"><div class="profile-wrapper text-center"><div id="avatar"> <a href="/" alt="avatar" class="mx-auto"> <img src="/assets/img/favicons/ms-icon-310x310.png" alt="avatar" onerror="this.style.display='none'"> </a></div><div class="site-title mt-3"> <a href="/">Hyo-Jun Lee</a></div><div class="site-subtitle font-italic">Bio-AI, Computer-Vision, Bio-Informatics</div></div><ul class="w-100"><li class="nav-item"> <a href="/" class="nav-link"> <i class="fa-fw fas fa-home ml-xl-3 mr-xl-3 unloaded"></i> <span>ABOUT ME</span> </a><li class="nav-item"> <a href="/tabs/publication/" class="nav-link"> <i class="fa-fw fas fa-archive ml-xl-3 mr-xl-3 unloaded"></i> <span>JOURNAL & CONFERENCE PUBLICATION</span> </a><li class="nav-item"> <a href="/tabs/categories/" class="nav-link"> <i class="fa-fw fas fa-stream ml-xl-3 mr-xl-3 unloaded"></i> <span>PROJECT</span> </a><li class="nav-item"> <a href="/tabs/tags/" class="nav-link"> <i class="fa-fw fas fa-tags ml-xl-3 mr-xl-3 unloaded"></i> <span>TAGS</span> </a></ul><div class="sidebar-bottom mt-auto d-flex flex-wrap justify-content-center"> <a href="https://github.com/gywns6287" aria-label="github" target="_blank" rel="noopener"> <i class="fab fa-github-alt"></i> </a> <a href=" javascript:location.href = 'mailto:' + ['gywns6287','gmail.com'].join('@')" aria-label="email" > <i class="fas fa-envelope"></i> </a></div></div><div id="topbar-wrapper" class="row justify-content-center topbar-down"><div id="topbar" class="col-11 d-flex h-100 align-items-center justify-content-between"> <span id="breadcrumb"> <span> <a href="/"> Posts </a> </span> <span>Numpy로 구현하는 GBLUP - Imprementing GBLUP with numpy</span> </span> <i id="sidebar-trigger" class="fas fa-bars fa-fw"></i><div id="topbar-title"> Post</div><i id="search-trigger" class="fas fa-search fa-fw"></i> <span id="search-wrapper" class="align-items-center"> <i class="fas fa-search fa-fw"></i> <input class="form-control" id="search-input" type="search" aria-label="search" placeholder="Search..."> <i class="fa fa-times-circle fa-fw" id="search-cleaner"></i> </span> <span id="search-cancel" >Cancel</span></div></div><div id="main-wrapper"><div id="main"><div class="row"><div id="post-wrapper" class="col-12 col-lg-11 col-xl-8"><div class="post pl-1 pr-1 pl-sm-2 pr-sm-2 pl-md-4 pr-md-4"><h1 data-toc-skip>Numpy로 구현하는 GBLUP - Imprementing GBLUP with numpy</h1><div class="post-meta text-muted d-flex flex-column"><div> <span class="timeago " data-toggle="tooltip" data-placement="bottom" title="Mon, Feb 1, 2021, 2:10 PM +0900" > Feb 1 <i class="unloaded">2021-02-01T14:10:00+09:00</i> </span> by <span class="author"> Hyo-Jun Lee </span></div><div> <span> Updated <span class="timeago lastmod" data-toggle="tooltip" data-placement="bottom" title="Wed, Mar 3, 2021, 2:41 PM +0900" > Mar 3 <i class="unloaded">2021-03-03T14:41:19+09:00</i> </span> </span> <span class="readtime" data-toggle="tooltip" data-placement="bottom" title="1960 words">10 min</span></div></div><div class="post-content"><p><small>사용된 데이터와 코드는 아래의 url 에서 확인 하실 수 있습니다. The data and codes used in the poster can be found at <strong>https://github.com/gywns6287/Implementing-GBLUP-with-numpy</strong> </small></p><p>이번 프로젝트에서는 Biointormatics에서 사용하는 genomic prediction 기법인 GBLUP에 대한 간략한 설명과 Numpy를 통한 구현방법을 소개해 드리고자 합니다.</p><p>In this poster, i will give a bried description of GBLUP and implement it by python Numpy.</p><h2 id="genomic-prediction">Genomic prediction??</h2><p>Genomic prediction 이란 생물의 유전적 정보로 부터 생물의 형질을 예측하는 기술입니다. 생물의 DNA, RNA 정보 등을 활용해 그 생명체의 형질을 미리 예측한다면, 아직 발생하지 않은 유전적 질병을 예측 하거나 아직 다 자라지 않은 개체가 성체가 되었을 때의 키, 몸무게 등을 예측 할 수 있을 것 입니다. 안젤리나 졸리가 BRCA 유전자 변이를 확인 하고 유방조직을 제거한 사례는 대표적인 Genomic prediction의 예시라고 할 수 있습니다. 마찬가지로 동물 분야에서는 가축의 예상 체중, 번식 능력 등을 예측하는데 Genomic prediction이 쓰일 수 있습니다.</p><p>Genomic prediction is a method that predicts the traits of an organism from it’s genetic information. If the traits of the living organism are predicted using the DNA and RNA information of the organism, it should be possible to predict a genetic disease that has not yet occurred, or predict the height and weight of an individual that has not yet grown up. The case where Angelina Jolie confirmed the BRCA gene mutation and removed her breast tissue is a representative example of genomic prediction. In a similar vein, in the animal field, genomic prediction can be used to predict the expected weight or reproductive capacity of livestock.</p><h2 id="gblup">GBLUP</h2><p>Genomic best linear unbiased prediction (GBLUP) 은 생물체의 양적 형질 예측에 가장 널리 쓰이는 Genomic prediction 기법입니다. GBLUP은 통계학에서 임의 효과(random effect)를 추정하기 위해 사용 되는 Best linear unbiased prediction (BLUP) 방법을 Genomic prediction에 적용시킨 기법입니다. GBLUP에서는 BLUP의 분산-공분산 행렬이 생물체 집단의 유전적 관계 행렬인 GRM으로 대체되어 있어 모델이 개체의 유전적 능력을 예측할 수 있게 됩니다. 다음 챕터 부터 순차적으로 DNA 정보로 부터 GRM 을 만들고 이를 BLUP 모델에 적용시켜 유전 능력을 예측하는 과정을 구현 해 보겠습니다.</p><p>Genomic best linear unbiased prediction (GBLUP) is the most widely used genomic prediction technique for quantitative trait prediction. GBLUP is applied the best linear unbiased prediction (BLUP) method, which is used to estimate random effects in statistics, to genomic prediction. It is an elegant method that simply uses a genomic relationship matrix (GRM) built from the genotypes instead of a variance-co-variance matrix in BLUP. From the next chapter, i will implement the process of calculating GRM from DNA information and applying them to the BLUP model for predicting genetic ability of individuals.</p><h2 id="calculate-grm">Calculate GRM</h2><p>이번 프로젝트에서는 Plink 의 raw 포맷의 데이터를 사용했습니다. 우선, 데이터를 불러옵니다.</p><p>In this poster, I used data in Plink’s raw format. First, load the data.</p><div class="language-python highlighter-rouge"><div class="highlight"><code><table class="rouge-table"><tbody><tr><td class="rouge-gutter gl"><pre class="lineno">1
2
3
4
5
</pre><td class="rouge-code"><pre><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="n">np</span>

<span class="c1">#Load Data
</span><span class="n">raw</span> <span class="o">=</span> <span class="s">'1100_sample.txt'</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">loadtxt</span><span class="p">(</span><span class="n">raw</span><span class="p">,</span> <span class="n">dtype</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="nb">str</span><span class="p">,</span> <span class="n">skiprows</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
</pre></table></code></div></div><hr /><p>데이터 형식을 살펴봅시다. 1번 컬럼에는 각 데이터의 참조 집단 여부가 표기되어 있습니다. G0으로 표기되어 있는 개체는 형질 정보를 알고 있는 참조집단이 되며 G1으로 표기되어 있는 개체는 형질 정보를 모르는 평가 집단 입니다. 2번 컬럼에는 개체의 이름이 표기 되어 있습니다. 3번과 4번 컬럼은 개체의 부모정보가 나타나 있습니다. 하지만 이번 프로젝트에서는 부모정보는 활용하지 않음으로 무시하셔도 좋습니다. 5번 컬럼에는 개체의 형질정보가 나타나 있으며 6번 이후의 컬럼에 개체의 DNA정보가 나타나 있습니다.</p><p>Let’s look at the data. Column 1 indicates each individual is included at reference population or not (G0: ref, G1: test). In column 2, the name of the individual is indicated. Columns 3 and 4 include the parent information of the individual. However, in this post, parental information is not used, so you can ignore it. In column 5, the individual’s trait information is included, and in the columns after number 6, the individual’s DNA information is included.</p><div class="language-python highlighter-rouge"><div class="highlight"><code><table class="rouge-table"><tbody><tr><td class="rouge-gutter gl"><pre class="lineno">1
2
3
4
5
6
7
</pre><td class="rouge-code"><pre><span class="k">print</span><span class="p">(</span><span class="n">data</span><span class="p">[:,:</span><span class="mi">8</span><span class="p">])</span>
<span class="k">print</span><span class="p">(</span><span class="s">'</span><span class="se">\n</span><span class="s">'</span><span class="p">,</span><span class="s">'#'</span><span class="o">*</span><span class="mi">60</span><span class="p">,</span><span class="s">'</span><span class="se">\n</span><span class="s">'</span><span class="p">)</span>

<span class="n">n_indiv</span><span class="p">,</span><span class="n">n_dna</span> <span class="o">=</span> <span class="n">data</span><span class="p">[:,</span><span class="mi">6</span><span class="p">:].</span><span class="n">shape</span>
<span class="k">print</span><span class="p">(</span><span class="s">'#N of indiv: {0}</span><span class="se">\n</span><span class="s">'</span>
<span class="s">'#N of DNA markers: {1}</span><span class="se">\n</span><span class="s">'</span>
<span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">n_indiv</span><span class="p">,</span><span class="n">n_dna</span><span class="p">))</span>
</pre></table></code></div></div><p>[[‘G0’ ‘11’ ‘0’ … ‘2.72785’ ‘0’ ‘2’]</p><p>[‘G0’ ‘15’ ‘0’ … ‘2.2779’ ‘0’ ‘1’]</p><p>[‘G0’ ‘17’ ‘0’ … ‘1.05285’ ‘0’ ‘1’]</p><p>…</p><p>[‘G1’ ‘19961’ ‘92’ … ‘-9’ ‘0’ ‘2’]</p><p>[‘G1’ ‘20067’ ‘254’ … ‘-9’ ‘0’ ‘0’]</p><p>[‘G1’ ‘20069’ ‘254’ … ‘-9’ ‘0’ ‘1’]]</p><p>############################################################ #N of indiv: 1100 #N of DNA markers: 49880</p><hr /><p>VanRaden (2008) 방법을 활용하여 GRM 을 계산해 봅시다. 우선 데이터 에서 dna 정보에 해당하는 컬럼 만을 가져 옵니다.</p><p>Let’s calculate the GRM using the VanRaden (2008) method. First, only the column corresponding to the DNA information is imported from the data.</p><div class="language-python highlighter-rouge"><div class="highlight"><code><table class="rouge-table"><tbody><tr><td class="rouge-gutter gl"><pre class="lineno">1
2
</pre><td class="rouge-code"><pre><span class="c1">#Separates dna information from raw files.
</span><span class="n">dna</span> <span class="o">=</span> <span class="n">data</span><span class="p">[:,</span><span class="mi">6</span><span class="p">:].</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="p">.</span><span class="nb">float</span><span class="p">)</span>
</pre></table></code></div></div><hr /><p>다음으로는 DNA 정보에 1을 빼주어 각 DNA 마커가 -1, 0, 1 로 보정 된 M 메트릭스를 만듭니다.</p><p>Next, subtract 1 from the DNA information to create an M matrix with each DNA marker corrected to -1, 0, 1.</p><div class="language-python highlighter-rouge"><div class="highlight"><code><table class="rouge-table"><tbody><tr><td class="rouge-gutter gl"><pre class="lineno">1
2
</pre><td class="rouge-code"><pre><span class="c1">#Calculate marker effects by genotype frequency
</span><span class="n">M</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">asmatrix</span><span class="p">(</span><span class="n">dna</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>
</pre></table></code></div></div><hr /><p>DNA genotype 효과를 계산한 P 메트릭스를 M 메트릭스에 빼줍니다. 이 과정은 더 희귀한 (rare) 한 DNA marker에 더 높은 값을 더하게 해줍니다.</p><p>Subtract the P matrix including the DNA genotype effect to the M matrix. This process allows M matrix to add a higher value for rare DNA marker.</p><div class="language-python highlighter-rouge"><div class="highlight"><code><table class="rouge-table"><tbody><tr><td class="rouge-gutter gl"><pre class="lineno">1
2
3
</pre><td class="rouge-code"><pre><span class="n">pi</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">mean</span><span class="p">(</span><span class="n">dna</span><span class="p">,</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span><span class="o">/</span><span class="mi">2</span>
<span class="n">P</span> <span class="o">=</span> <span class="mi">2</span><span class="o">*</span><span class="p">(</span><span class="n">pi</span><span class="o">-</span><span class="mf">0.5</span><span class="p">)</span>
<span class="n">M</span> <span class="o">=</span> <span class="n">M</span> <span class="o">-</span> <span class="n">P</span>
</pre></table></code></div></div><hr /><p>VanRaden (2008)이 명시한 수식 을 활용해 G를 계산합니다.</p><p>G is calculated using the formula described at VanRaden, 2008.</p><p>\(G=\frac{MM'}{2\sum p_{i}(1-p_{i})}\)</p><pre><code class="language-pyhon">#Calculate G by VanRaden method
G = M*M.T/(2*np.sum(pi*(1-pi)))
</code></pre><h2 id="calculate-gebv">Calculate gEBV</h2><p>동물의 유전능력을 평가에서는 GBLUP을 gEBV (genomic estimated breeding value) 을 계산하기 위해 활용합니다. GBLUP 을 통해 gEBV를 계산하는 공식은 다음과 같습니다.</p>\[gEBV=(Z^{T}Z+G^{-1})^{-1}Z^{T}(y-\hat{y})\]<p>여기서 \(Z\)는 임의효과를 위한 designed matrix (자세한 설명은 다음 쳅터에), \(y\)는 개체의 형질, \(\hat{y}\)는 형질의 평균을 의미합니다.</p><p>In evaluating the genetic capacity of animals, GBLUP is used to calculate the genomic estimated breeding value (gEBV). The formula for calculating gEBV is shown above. \(Z\) is a designed matrix for random effects (detailed in the next chapter), \(y\) is the trait of the individual, and \(\hat{y}\) is the average of the traits.</p><p>위의 수식을 구현하기위해 우선 형질의 편차 (\(y-\hat{y}\)) 를 계산합니다.</p><p>To implement the above equation, calculate the deviation of the trait (\(y-\hat{y}\)) .</p><div class="language-python highlighter-rouge"><div class="highlight"><code><table class="rouge-table"><tbody><tr><td class="rouge-gutter gl"><pre class="lineno">1
2
3
</pre><td class="rouge-code"><pre><span class="c1">#Calculate the deviation of trait(y)
</span><span class="n">y</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">data</span><span class="p">[:,</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s">'G0'</span><span class="p">,</span><span class="mi">5</span><span class="p">].</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="p">.</span><span class="nb">float</span><span class="p">)</span>
<span class="n">y_c</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">asmatrix</span><span class="p">(</span><span class="n">y</span> <span class="o">-</span> <span class="n">np</span><span class="p">.</span><span class="n">mean</span><span class="p">(</span><span class="n">y</span><span class="p">)).</span><span class="n">T</span>
</pre></table></code></div></div><hr /><p>다음으로 임의효과를 위한 디자인 메트릭스 (\(Z\)) 를 만듭니다. 디자인 메트릭스는 (참조집단의 수\(\times\)전체 집단의 수) 의 구조를 띄고 있습니다. 참조집단에 해당하는 열은 같은 수의 행에 1을 갖고, 평가 집단에 해당하는 열은 모두 0 으로 채워 줍니다. 설명이 복잡하다면 아래의 코드를 참고하시면 됩니다.</p><p>Next, create a design matrix (\(Z\)) for random effects. Design metrics have a structure of (number of reference \(\times\) number of total population). The column corresponding to the reference has 1 in the same number of rows, and columns corresponding to the evaluation group are filled with zeros. If the explanation is complicated, you can refer to the code below.</p><div class="language-python highlighter-rouge"><div class="highlight"><code><table class="rouge-table"><tbody><tr><td class="rouge-gutter gl"><pre class="lineno">1
2
3
4
5
6
7
8
9
</pre><td class="rouge-code"><pre><span class="c1">#Make Zeros matrix
</span><span class="n">Z</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">sum</span><span class="p">(</span><span class="n">data</span><span class="p">[:,</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s">'G0'</span><span class="p">),</span><span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)))</span>

<span class="c1">#Fill 1 at the same row and column for reference population.  
</span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span>  <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)):</span>
    <span class="k">if</span> <span class="n">data</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s">'G0'</span><span class="p">:</span>
        <span class="n">Z</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>

<span class="n">Z</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">asmatrix</span><span class="p">(</span><span class="n">Z</span><span class="p">)</span>
</pre></table></code></div></div><hr /><p>앞서 설명한 공식을 구현하여 gEBV를 계산합니다.</p><p>Calculate gEBV by implementing the previously described formula.</p><div class="language-python highlighter-rouge"><div class="highlight"><code><table class="rouge-table"><tbody><tr><td class="rouge-gutter gl"><pre class="lineno">1
2
3
4
5
6
</pre><td class="rouge-code"><pre><span class="c1">#calculate G inverse
</span><span class="n">G_inv</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">linalg</span><span class="p">.</span><span class="n">inv</span><span class="p">(</span><span class="n">G</span><span class="p">)</span>

<span class="c1">#estimate gebv
</span><span class="n">gebv</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">linalg</span><span class="p">.</span><span class="n">inv</span><span class="p">(</span><span class="n">Z</span><span class="p">.</span><span class="n">T</span> <span class="o">*</span> <span class="n">Z</span> <span class="o">+</span> <span class="n">G_inv</span><span class="p">)</span> <span class="o">*</span> <span class="n">Z</span><span class="p">.</span><span class="n">T</span> <span class="o">*</span> <span class="n">y_c</span>
<span class="k">print</span><span class="p">(</span><span class="n">gebv</span><span class="p">)</span>
</pre></table></code></div></div><p>[[-0.10661197]</p><p>[-0.37285056]</p><p>[-0.97929329]</p><p>…</p><p>[-0.37161146]</p><p>[-0.01737475]</p><p>[ 0.04712397]]</p><h2 id="reference">Reference</h2><p>VanRaden, Paul M. “Efficient methods to compute genomic predictions.” <em>Journal of dairy science</em> 91.11 (2008): 4414-4423.</p></div><div class="post-tail-wrapper text-muted"><div class="post-meta mb-3"> <i class="far fa-folder-open fa-fw mr-1"></i> <a href='/categories/bio-informatics/'>Bio-Informatics</a></div><div class="post-tags"> <i class="fa fa-tags fa-fw mr-1"></i> <a href="/tags/genomic-prediction/" class="post-tag no-text-decoration" >Genomic-Prediction</a> <a href="/tags/animal-science/" class="post-tag no-text-decoration" >Animal-Science</a></div><div class="post-tail-bottom d-flex justify-content-between align-items-center mt-3 pt-5 pb-2"><div class="license-wrapper"> This post is licensed under <a href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a> by the author.</div><div class="share-wrapper"> <span class="share-label text-muted mr-1">Share</span> <span class="share-icons"> <a href="https://twitter.com/intent/tweet?text=Numpy로 구현하는 GBLUP - Imprementing GBLUP with numpy - Hyo-Jun Lee&url=https://gywns6287.github.io/posts/GBLUP/" data-toggle="tooltip" data-placement="top" title="Twitter" target="_blank" rel="noopener" aria-label="Twitter"> <i class="fa-fw fab fa-twitter"></i> </a> <a href="https://www.facebook.com/sharer/sharer.php?title=Numpy로 구현하는 GBLUP - Imprementing GBLUP with numpy - Hyo-Jun Lee&u=https://gywns6287.github.io/posts/GBLUP/" data-toggle="tooltip" data-placement="top" title="Facebook" target="_blank" rel="noopener" aria-label="Facebook"> <i class="fa-fw fab fa-facebook-square"></i> </a> <i class="fa-fw fas fa-link small" onclick="copyLink()" data-toggle="tooltip" data-placement="top" title="Copy link"></i> </span></div></div></div></div></div><div id="panel-wrapper" class="col-xl-3 pl-2 text-muted topbar-down"><div class="access"><div id="access-lastmod" class="post"> <span>Recent Update</span><ul class="post-content pl-0 pb-1 ml-1 mt-2"><li><a href="/posts/PIGSE/">ResNet50 을 활용한 돼지 도체의 성별 예측</a><li><a href="/posts/GBLUP/">Numpy로 구현하는 GBLUP - Imprementing GBLUP with numpy</a></ul></div><div id="access-tags"> <span>Trending Tags</span><div class="d-flex flex-wrap mt-3 mb-1 mr-3"> <a class="post-tag" href="/tags/animal-science/">Animal-Science</a> <a class="post-tag" href="/tags/computer-vision/">Computer-Vision</a> <a class="post-tag" href="/tags/genomic-prediction/">Genomic-Prediction</a></div></div></div><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@1.0.1/dist/bootstrap-toc.min.js"></script><div id="toc-wrapper" class="pl-0 pr-4 mb-5"> <span class="pl-3 pt-2 mb-2">Contents</span><nav id="toc" data-toggle="toc"></nav></div></div></div><div class="row"><div class="col-12 col-lg-11 col-xl-8"><div id="post-extend-wrapper" class="pl-1 pr-1 pl-sm-2 pr-sm-2 pl-md-4 pr-md-4"><div id="related-posts" class="mt-5 mb-2 mb-sm-4"><h3 class="pt-2 mt-1 mb-4 ml-1" data-toc-skip>Further Reading</h3><div class="card-deck mb-4"><div class="card"> <a href="/posts/PIGSE/"><div class="card-body"> <span class="timeago small" > Mar 2 <i class="unloaded">2021-03-02T17:00:00+09:00</i> </span><h3 class="pt-0 mt-1 mb-3" data-toc-skip>ResNet50 을 활용한 돼지 도체의 성별 예측</h3><div class="text-muted small"><p> 사용된 코드는 아래의 url 에서 확인 하실 수 있습니다. The codes used in the poster can be found at https://github.com/gywns6287/pig_sex_classifier 이번 프로젝트에서는 컴퓨터 비전 영역의 성공한 딥러닝 모델인 ResNet50 를 활용하여 돼지 도체의 성별을 예측하는 모델을...</p></div></div></a></div></div></div><div class="post-navigation d-flex justify-content-between"> <span class="btn btn-outline-primary disabled"><p>-</p></span> <a href="/posts/PIGSE/" class="btn btn-outline-primary" prompt="Newer"><p>ResNet50 을 활용한 돼지 도체의 성별 예측</p></a></div><div id="disqus" class="pt-2 pb-2"><p class="text-center text-muted pb-5"> Loading comments from <a href="https://disqus.com/">Disqus</a> ...</p></div><script src="/assets/js/lib/jquery.disqusloader.min.js"></script> <script> const options = { scriptUrl: '//gywns6287.disqus.com/embed.js', disqusConfig: function() { this.page.title = 'Numpy로 구현하는 GBLUP - Imprementing GBLUP with numpy'; this.page.url = 'https://gywns6287.github.io/posts/GBLUP/'; this.page.identifier = '/posts/GBLUP/'; } }; $.disqusLoader('#disqus', options); </script></div></div></div><footer class="d-flex w-100 justify-content-center"><div class="d-flex justify-content-between align-items-center"><div class="footer-left"><p class="mb-0"> © 2021 <a href="https://twitter.com/username">Hyo-Jun Lee</a>. <span data-toggle="tooltip" data-placement="top" title="Except where otherwise noted, the blog posts on this site are licensed under the Creative Commons Attribution 4.0 International (CC BY 4.0) License by the author.">Some rights reserved.</span></p></div><div class="footer-right"><p class="mb-0"> Powered by <a href="https://jekyllrb.com" target="_blank" rel="noopener">Jekyll</a> with <a href="https://github.com/cotes2020/jekyll-theme-chirpy" target="_blank" rel="noopener">Chirpy</a> theme.</p></div></div></footer></div><div id="search-result-wrapper" class="d-flex justify-content-center unloaded"><div class="col-12 col-xl-11 post-content"><div id="search-hints"><h4 class="text-muted mb-4">Trending Tags</h4><a class="post-tag" href="/tags/animal-science/">Animal Science</a> <a class="post-tag" href="/tags/computer-vision/">Computer Vision</a> <a class="post-tag" href="/tags/genomic-prediction/">Genomic Prediction</a></div><div id="search-results" class="d-flex flex-wrap justify-content-center text-muted mt-3"></div></div></div></div><div id="mask"></div><a id="back-to-top" href="#" aria-label="back-to-top" class="btn btn-lg btn-box-shadow" role="button"> <i class="fas fa-angle-up"></i> </a> <script src="https://cdn.jsdelivr.net/npm/simple-jekyll-search@1.7.3/dest/simple-jekyll-search.min.js"></script> <script> SimpleJekyllSearch({ searchInput: document.getElementById('search-input'), resultsContainer: document.getElementById('search-results'), json: '/assets/js/data/search.json', searchResultTemplate: '<div class="pl-1 pr-1 pl-sm-2 pr-sm-2 pl-lg-4 pr-lg-4 pl-xl-0 pr-xl-0"> <a href="https://gywns6287.github.io{url}">{title}</a><div class="post-meta d-flex flex-column flex-sm-row text-muted mt-1 mb-1"><div class="mr-sm-4"><i class="far fa-folder fa-fw"></i>{categories}</div><div><i class="fa fa-tag fa-fw"></i>{tags}</div></div><p>{snippet}</p></div>', noResultsText: '<p class="mt-5">Oops! No result founds.</p>' }); </script> <script defer src="https://www.googletagmanager.com/gtag/js?id=G-SLE093F45F"></script> <script> document.addEventListener("DOMContentLoaded", function(event) { window.dataLayer = window.dataLayer || []; function gtag(){dataLayer.push(arguments);} gtag('js', new Date()); gtag('config', 'G-SLE093F45F'); }); </script>
